_source/_posts/2021-11-08-k8s-api-server-oidc.md

```diff
@@ -0,0 +1,561 @@
+  "nonce": "2Rsv7E84vo_K14J6qm9IgKjNsCUxb4Hc16qgzx9CKlk",
+  "auth_time": 1634649927,
+  "at_hash": "veXQkvqTjCl70OkZnKZVyA",
+  "groups": [
+    "k8s-admins"
+  ]
+}
+
+```
+
+Now let us update kubectl configuration to add an OIDC user:
+
+```bash
+kubectl config set-credentials oidc-user \
+--exec-api-version=client.authentication.k8s.io/v1beta1 \
+--exec-command=kubectl \
+--exec-arg=oidc-login \
+--exec-arg=get-token \
+--exec-arg=--oidc-issuer-url=<k8s_oidc_issuer_url> \
+--exec-arg=--oidc-client-id=<k8s_oidc_client_id> \
+--exec-arg=--oidc-extra-scope="email offline_access profile openid"
+```
+
+Now we can use this user to authenticate to the cluster via kubectl.
+
+```bash
+kubectl get pods --user=oidc-user -n default
+```
+
+This should open a browser window to authenticate you. You can use your Okta account credentials to log in.
+
+You can play around with different users in different groups to test out the RBAC configuration we did. Users who are not in the `k8s-admins` group in Okta should not be able to access any resource. Users who are only in the `restricted-users` group should only be able to see pods and services.
+
+You can use the below command to set this user as default for your current kubectl context.
+
+```
+kubectl config set-context --current --user=oidc-user
+```
+
+# Learn more about using OIDC with Kubernetes
+
+Using OIDC is a great way to secure your Kubernetes clusters, especially in bigger teams. It is more secure than any of the default Kubernetes authentication mechanisms. On top of that, it will let you manage users and their roles and permissions in your cluster and even add multi-factor authentication for your cluster.
+
+While this exercise shows how to do it with Okta, the process would be very similar for other OIDC providers as well. Check out the [Okta Terraform provider docs](https://registry.terraform.io/providers/okta/okta/latest/docs) to see what else can be automated via Terraform. You can also automate the Kubernetes parts via terraform using the [Kubernetes Terraform provider](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs).
+
+If you want to learn more about Kubernetes, OIDC, or using OIDC with Kubernetes, check out these additional resources.
+
+- [OAuth 2.0 and OpenID Connect Overview](https://developer.okta.com/docs/concepts/oauth-openid/)
+- [Secure Access to AWS EKS Clusters for Admins](/blog/2021/10/08/secure-access-to-aws-eks)
+- [Build a Microservice Architecture with Spring Boot and Kubernetes](/blog/2019/04/01/spring-boot-microservices-with-kubernetes)
+- [Kubernetes to the Cloud with Spring Boot and JHipster](/blog/2021/06/01/kubernetes-spring-boot-jhipster)
+- [Using Okta Advanced Server Access & Terraform to Automate Identity & Infrastructure as Code](/blog/2020/04/24/okta-terraform-automate-identity-and-infrastructure)
+
+You can find the code for this tutorial on GitHub at <https://github.com/oktadev/okta-k8s-oidc-terraform-example>.
```

Comment:
I just re-examined this repo after polishing up a contractor one. I noticed it has an [MIT license](https://github.com/oktadev/okta-k8s-oidc-terraform-example/blob/main/LICENSE) instead of Apache. We use Apache 2.0 on all our examples and add Okta, Inc. as the copyright holder.

Related: please fill out the README so if users stumble upon your repo first, they can get it up and running w/o reading the post. Or, just add some text that encourages them to read the post. Here's the example repo I updated this morning:

https://github.com/oktadev/okta-java-records-example

Reply:
Hey thank you I'll fix them. These are the kind of newbie mistakes I was afraid of

Reply:
I updated these. Added instruction on the Readme to use the repo standalone and pointed reader toward blog for further setup